core/src/main/java/org/apache/iceberg/avro/Avro.java

```diff
@@ -321,6 +321,8 @@ public DataWriteBuilder withSortOrder(SortOrder newSortOrder) {
 
     public <T> DataWriter<T> build() throws IOException {
       Preconditions.checkArgument(spec != null, "Cannot create data writer without spec");
+      Preconditions.checkArgument(spec.isUnpartitioned() || partition != null,
```

Comment:
Extra validation I originally had in the write factory impl but I think it makes sense to move it here.